### PR TITLE
feat(logging): migration to structure logging for main.go, pkg/ & internal/ folders

### PR DESCRIPTION
**Pull request by dmpe**
**What this PR does / why we need it**:

/wg structured-logging 
/area logging
/priority important-longterm
/kind cleanup
/cc @kubernetes/wg-structured-logging-reviews

**How does this change affect the cardinality of KSM**: *(increases, decreases or does not change cardinality)*

does not change cardinality

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

part 1 of adding JSON support https://github.com/kubernetes/kube-state-metrics/issues/1485, i.e. with this PR we introduce structure logging, by following the associated KEP  https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md

### CODE REVIEW

File: pkg/customresourcestate/registry_factory.go [hunk 1]
```diff
@@ -474,7 +474,7 @@ func generate(u *unstructured.Unstructured, f compiledFamily, errLog klog.Verbos
 		v.DefaultLabels(baseLabels)
 		metrics = append(metrics, v.ToMetric())
 	}
-	klog.V(10).Infof("%s: produced %d metrics for %s", f.Name, len(metrics), u.GetName())
+	klog.V(10).InfoS("produced metrics for", "compiledFamilyName", f.Name, "metricsLength", len(metrics), "unstructuredName", u.GetName())
```
Comment (dmpe) [inline pkg/customresourcestate/registry_factory.go:477]:
```suggestion
	klog.V(10).InfoS("Produced metrics for", "compiledFamilyName", f.Name, "metricsLength", len(metrics), "unstructuredName", u.GetName())
```

File: pkg/customresourcestate/registry_factory.go [hunk 2]
```diff
@@ -461,7 +461,7 @@ func famGen(f compiledFamily) generator.FamilyGenerator {
 }
 
 func generate(u *unstructured.Unstructured, f compiledFamily, errLog klog.Verbose) *metric.Family {
-	klog.V(10).Infof("%s: checking %s", f.Name, u.GetName())
+	klog.V(10).InfoS("checking", "compiledFamilyName", f.Name, "unstructuredName", u.GetName())
```
Comment (dmpe) [inline pkg/customresourcestate/registry_factory.go:464]:
```suggestion
	klog.V(10).InfoS("Checked", "compiledFamilyName", f.Name, "unstructuredName", u.GetName())
```

**Comment by dmpe** [comment]
I have tested this locally via e2e tests and it did not fail :) 

When starting kube state metrics before my PR:

```
 ./kube-state-metrics 
I0813 14:05:15.320475   51084 server.go:93] Using default resources
I0813 14:05:15.322843   51084 types.go:136] Using all namespace
I0813 14:05:15.322881   51084 server.go:122] Metric allow-denylisting: Excluding the following lists that were on denylist: 
W0813 14:05:15.322905   51084 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
W0813 14:05:15.322917   51084 client_config.go:622] error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
F0813 14:05:15.324143   51084 main.go:65] Failed to run kube-state-metrics: failed to create client: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
goroutine 1 [running]:
k8s.io/klog/v2.stacks(0x1)
	/home/jm/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:860 +0x8a
k8s.io/klog/v2.(*loggingT).output(0x265b260, 0x3, 0x0, 0xc000421960, 0x1, {0x1e675c7?, 0x1?}, 0x265c460?, 0x0)
	/home/jm/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:825 +0x686
k8s.io/klog/v2.(*loggingT).printfDepth(0x265b260, 0x0?, 0x0, {0x0, 0x0}, 0x7ffa5af073c0?, {0x18a8a12, 0x24}, {0xc0003ba8c0, 0x1, ...})
	/home/jm/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:630 +0x1f2
k8s.io/klog/v2.(*loggingT).printf(...)
	/home/jm/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:612
k8s.io/klog/v2.Fatalf(...)
	/home/jm/go/pkg/mod/k8s.io/klog/v2@v2.60.1/klog.go:1516
main.main()
	/home/jm/Documents/kube-state-metrics/main.go:65 +0x514
```

And after my proposed changes.

```
./kube-state-metrics 
I0813 14:08:53.765129   51528 server.go:93] "Used default resources"
I0813 14:08:53.765297   51528 types.go:136] "Using all namespace"
I0813 14:08:53.765314   51528 server.go:122] "Metric allow-denylisting" allowDenyStatus="Excluding the following lists that were on denylist: "
W0813 14:08:53.765342   51528 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
W0813 14:08:53.765351   51528 client_config.go:622] error creating inClusterConfig, falling back to default config: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
E0813 14:08:53.765443   51528 main.go:67] "Failed to run kube-state-metrics" err="failed to create client: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable"
```

File: pkg/metricshandler/metrics_handler.go [hunk 3]
```diff
@@ -93,14 +93,14 @@ func (m *MetricsHandler) Run(ctx context.Context) error {
...
-		klog.Info("Autosharding disabled")
+		klog.InfoS("Autosharding disabled")
 		m.ConfigureSharding(ctx, m.opts.Shard, m.opts.TotalShards)
 		<-ctx.Done()
 		return ctx.Err()
 	}
 
-	klog.Infof("Autosharding enabled with pod=%v pod_namespace=%v", m.opts.Pod, m.opts.Namespace)
-	klog.Infof("Auto detecting sharding settings.")
+	klog.InfoS("Autosharding enabled with pod pod_namespace", "pod", m.opts.Pod, "namespace", m.opts.Namespace)
```
Comment (yangjunmyfm192085) [inline pkg/metricshandler/metrics_handler.go:102]:
```suggestion
	klog.InfoS("Autosharding enabled with pod", "pod", klog.KRef(m.opts.Namespace, m.opts.Pod))
```

File: pkg/options/types.go [hunk 4]
```diff
@@ -133,13 +133,13 @@ func (n *NamespaceList) Set(value string) error {
 func (n *NamespaceList) GetNamespaces() NamespaceList {
 	ns := *n
 	if len(*n) == 0 {
-		klog.Info("Using all namespace")
+		klog.InfoS("Using all namespace")
```
Comment (yangjunmyfm192085) [inline pkg/options/types.go:136]:
```suggestion
		klog.InfoS("Using all namespaces")
```

File: pkg/options/types.go [hunk 5]
```diff
@@ -133,13 +133,13 @@ func (n *NamespaceList) Set(value string) error {
...
 		ns = DefaultNamespaces
 	} else {
 		if n.IsAllNamespaces() {
-			klog.Info("Using all namespace")
+			klog.InfoS("Using all namespace")
```
Comment (yangjunmyfm192085) [inline pkg/options/types.go:140]:
```suggestion
			klog.InfoS("Using all namespaces")
```

File: pkg/metricshandler/metrics_handler.go (hunk 3 above)
Comment (dmpe) [reply pkg/metricshandler/metrics_handler.go:102]: thank you! I was reading about that, but still missed....

File: pkg/options/types.go [hunk 6]
```diff
@@ -133,13 +133,13 @@ func (n *NamespaceList) Set(value string) error {
...
 		} else {
-			klog.Infof("Using %s namespaces", ns)
+			klog.InfoS("Using namespaces", "namespaces", ns)
```
Comment (yangjunmyfm192085) [inline pkg/options/types.go:142]:
```suggestion
			klog.InfoS("Using namespaces", "nameSpaces", ns)
```

**Review by mrueg — APPROVED** [review]
Thanks!
/lgtm

/hold 
for @dgrisonnet @fpetkovski to review.

File: internal/store/builder.go [hunk 7]
```diff
@@ -178,7 +178,7 @@ func (b *Builder) WithCustomResourceStoreFactories(fs ...customresource.Registry
 	for i := range fs {
 		f := fs[i]
 		if _, ok := availableStores[f.Name()]; ok {
-			klog.Warningf("The internal resource store named %s already exists and is overridden by a custom resource store with the same name, please make sure it meets your expectation", f.Name())
```
Comment (fpetkovski) [inline internal/store/builder.go:181]: Is this because we don't have a way to log warn messages with structure logging?

Comment (mrueg) [reply internal/store/builder.go:181]: Yes, that does not exist there. See also: https://github.com/kubernetes/klog/pull/184

**Pull request merged**